In [334]:
import pandas as pd

# 데이터 불러오기
url = 'https://github.com/MyungKyuYi/AI-class/raw/main/titanic.csv'
data = pd.read_csv(url)

In [336]:
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [338]:
print("변환 전 레이블 분포:")
print(data['Survived'].value_counts())

변환 전 레이블 분포:
Survived
0    549
1    342
Name: count, dtype: int64


In [340]:
data.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [342]:
missing_values = data.isnull().sum()
print(missing_values)

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


In [344]:
data['Age'] = data['Age'].fillna(data['Age'].mean())

888의 age의 nan이 age의 평균값으로 대체

data['Age'].mean() : 평균

fillna() : 대체

In [349]:
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex        Age  \
0                              Braund, Mr. Owen Harris    male  22.000000   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.000000   
2                               Heikkinen, Miss. Laina  female  26.000000   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.000000   
4                             Allen, Mr. William Henry    male  35.000000   
..                                                 ...     ...        ...   
886                              Montvila, Rev. Juozas    male  27.000000   
887                       Graham, Miss. Margaret Edith  female  19.000000   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.699118   
889                              Behr, Mr. Karl Howell    male  26.000000   
890                                Dooley, Mr. Patrick    male  32.000000   

     SibSp  Parch            Ticket     Fare Cabin Embarked  
0        1      0         A/5 21171   7.2500   NaN        S  
1        1      0          PC 17599  71.2833   C85        C  
2        0      0  STON/O2. 3101282   7.9250   NaN        S  
3        1      0            113803  53.1000  C123        S  
4        0      0            373450   8.0500   NaN        S  
..     ...    ...               ...      ...   ...      ...  
886      0      0            211536  13.0000   NaN        S  
887      0      0            112053  30.0000   B42        S  
888      1      2        W./C. 6607  23.4500   NaN        S  
889      0      0            111369  30.0000  C148        C  
890      0      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

불필요한 칼럼제거(생존률과 관련이 없는 name, ticket cabin, embarked)

drop 행 또는 열 삭제, dropna 결측치가 포함된 행 또는 열 삭제

In [353]:
data.drop(columns=['Name', 'Ticket', 'Cabin', 'Embarked'], inplace=True,errors='ignore')

In [355]:
data

PassengerId  Survived  Pclass     Sex        Age  SibSp  Parch     Fare
0              1         0       3    male  22.000000      1      0   7.2500
1              2         1       1  female  38.000000      1      0  71.2833
2              3         1       3  female  26.000000      0      0   7.9250
3              4         1       1  female  35.000000      1      0  53.1000
4              5         0       3    male  35.000000      0      0   8.0500
..           ...       ...     ...     ...        ...    ...    ...      ...
886          887         0       2    male  27.000000      0      0  13.0000
887          888         1       1  female  19.000000      0      0  30.0000
888          889         0       3  female  29.699118      1      2  23.4500
889          890         1       1    male  26.000000      0      0  30.0000
890          891         0       3    male  32.000000      0      0   7.7500

[891 rows x 8 columns]

In [357]:
data['Sex'] = data['Sex'].map({'male': 0, 'female': 1})

In [359]:
data

PassengerId  Survived  Pclass  Sex        Age  SibSp  Parch     Fare
0              1         0       3    0  22.000000      1      0   7.2500
1              2         1       1    1  38.000000      1      0  71.2833
2              3         1       3    1  26.000000      0      0   7.9250
3              4         1       1    1  35.000000      1      0  53.1000
4              5         0       3    0  35.000000      0      0   8.0500
..           ...       ...     ...  ...        ...    ...    ...      ...
886          887         0       2    0  27.000000      0      0  13.0000
887          888         1       1    1  19.000000      0      0  30.0000
888          889         0       3    1  29.699118      1      2  23.4500
889          890         1       1    0  26.000000      0      0  30.0000
890          891         0       3    0  32.000000      0      0   7.7500

[891 rows x 8 columns]

엔코딩 : 문자형 데이터를 머신러닝 모델이 이해할 수 있게 숫자로 바꿔주는 것

In [362]:
print(data['Survived'].value_counts())

Survived
0    549
1    342
Name: count, dtype: int64


In [364]:
from imblearn.over_sampling import SMOTE

X = data.drop(columns=['Survived'])  # 피처(입력값)
y = data['Survived']  # 레이블(정답)

smote = SMOTE(sampling_strategy='auto', random_state=42)
X_resampled, y_resampled = smote.fit_resample(X, y)

# 변환 후 레이블 개수 확인
print("변환 후 레이블 분포:")
print(y_resampled.value_counts())

변환 후 레이블 분포:
Survived
0    549
1    549
Name: count, dtype: int64


데이터 불균형을 해결하기 위해 오버샘플링(SMOTE)

In [367]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix

# 데이터 준비
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.2, random_state=42)

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# 모델 리스트
models = {
    "Random Forest": RandomForestClassifier(),
    "Decision Tree": DecisionTreeClassifier(),
    "Logistic Regression": LogisticRegression(max_iter=5000),
    "KNN": KNeighborsClassifier(),
    "SVM": SVC()
}

# 모델 학습 및 평가
for name, model in models.items():
    model.fit(X_train, y_train)  # 모델 학습
    y_pred = model.predict(X_test)  # 예측
    acc = accuracy_score(y_test, y_pred)  # 정확도
    cm = confusion_matrix(y_test, y_pred)  # 혼동 행렬
    
    print(f"\n {name} 성능 평가:")
    print(f"정확도(Accuracy): {acc:.4f}")
    print("Confusion Matrix:")
    print(cm)


 Random Forest 성능 평가:
정확도(Accuracy): 0.8273
Confusion Matrix:
[[87 16]
 [22 95]]

 Decision Tree 성능 평가:
정확도(Accuracy): 0.7682
Confusion Matrix:
[[83 20]
 [31 86]]

 Logistic Regression 성능 평가:
정확도(Accuracy): 0.7955
Confusion Matrix:
[[80 23]
 [22 95]]

 KNN 성능 평가:
정확도(Accuracy): 0.6591
Confusion Matrix:
[[67 36]
 [39 78]]

 SVM 성능 평가:
정확도(Accuracy): 0.6182
Confusion Matrix:
[[100   3]
 [ 81  36]]


로지스틱 회귀 수렴 문제 : 로지스틱 회귀(Logistic Regression) 모델이 최적 해(수렴점)에 도달하지 못해 정규화와 반복횟수(max_iter) 설정